waterbutler/providers/onedrive/provider.py

```
@@ -0,0 +1,419 @@
+
+        resp = await self.make_request(
+            'GET',
+            url,
+            expects=(200, 404, ),
+            throws=exceptions.ProviderError
+        )
+
+        if (resp.status == 404):
+            ids = None
+            folder = False
+            await resp.release()
+        else:
+            data = await resp.json()
+            ids = self._get_ids(data)[-1]
+            folder = ('folder' in data.keys())
+
+        return base.child(path, _id=ids, folder=folder)
+
+    @property
+    def default_headers(self):
+        return {
+            'Authorization': 'Bearer {}'.format(self.token),
+        }
+
+    async def intra_copy(self, dest_provider, src_path, dest_path):
+        """ OneDrive API Reference: https://dev.onedrive.com/items/copy.htm """
+
+        url = self.build_url(src_path.identifier, 'action.copy')
+        payload = json.dumps({'name': dest_path.name,
+                              'parentReference': {'id': dest_path.parent.full_path.strip('/') if dest_path.parent.identifier is None else dest_path.parent.identifier}})  # TODO: this feels like a hack.  parent.identifier is None in some cases.
+
+        logger.info('intra_copy dest_provider::{} src_path::{} dest_path::{}  url::{} payload::{}'.format(repr(dest_provider), repr(src_path), repr(dest_path), repr(url), payload))
+        resp = await self.make_request(
+            'POST',
+            url,
+            data=payload,
+            headers={'content-type': 'application/json', 'Prefer': 'respond-async'},
+            expects=(202, ),
+            throws=exceptions.IntraCopyError,
+        )
+        if resp is None:
+            raise exceptions.IntraCopyError
+        logger.info('resp::{}'.format(repr(resp)))
+        status_url = resp.headers['LOCATION']
+        await resp.release()
+        logger.info('status_url::{}'.format(repr(status_url)))
+        i = 0
+        status = None
+        while (i < 30):
```

@felliott: I added count and sleep interval to the settings file, good point/idea.  As far as what happens if it fails, I would think it should throw an exception since that should be an exceptional / unusual event.  

What would be a "good" exception and text/language to use?  Something like: OneDrive file copy failed.  Please wait for 1-2 minutes and reload the page to see if OneDrive was able to copy the file.